doc/sphinx/src/using-eos.rst

```diff
@@ -3,3 +3,493 @@
+
+  eos.Finalize();
+
+Vector and Scalar API, Accessors
+---------------------------------
+
+Most ``EOS`` methods have both scalar and vector overloads, where the
+scalar version returns a value, and the vector version modifies an
+array. By default the vector version is called from host on device (if
+``singularity-eos`` was compiled for device).
+
+The vector API is templated to accept *accessors*. An accessor is any
+object with a square bracket operator. One-dimensional arrays,
+pointers, and ``std::vector<double>`` are two exmaples of what we call
+an accessor. However, the value of an accessor is it doesn't have to
+be an array. You can create an accessor class that wraps your
+preferred memory layout, and ``singularity-eos`` will handle it
+appropriately. An accessor that indexes into an array with some stride
+might look like this:
+
+.. code-block:: cpp
+
+  struct Indexer {
+    Indexer(int stride, double *array) : stride_(stride), A_(array) {}
+    double &operator[](int i) {
+      return A_[stride*i];
+    }
+    double *A_;
+    int stride_;
+  };
+
+We do note, however, that vectorization may suffer if your underlying
+data structure is not contiguous in memory.
+
+Lambdas and Optional Parameters
+--------------------------------
+
+Most methods for ``EOS`` objects accept an optional ``lambda``
+parameter, which is a ``Real *``. Unless specified in :ref:`the
+models section <models>`, this parameter does nothing. However, some
+models require or benefit from additional information. For example
+models with internal root finds can leverage initial guesses and
+models with composition mixing parameters may need additional input to
+return a meaningful state.
+
+``EOS`` models are introspective and can provide the desired/required
+size of the lambda array with:
+
+.. cpp:function:: int EOS::nlambda()
+
+which is the desired size of the ``lambda`` array per scalar call. For
+vector calls, there should be one such array per grid point. An
+accessor for ``lambda`` should return a ``Real *`` pointer at each
+index. A trivial example of such an indexer for ``lambda`` might be
+the null indexer:
+
+.. code-block:: cpp
+
+  class NullIndexer {
+    Real *operator[](int i) { return nullptr; }
+  };
+
+As a general rule, to avoid race conditions, you will want at least
+one ``lambda`` array (or subview of a larger memory allocation) per
+thread. You may want one array per point you are evaluating
+on. Ideally these arrays are persistent between ``EOS`` calls, to
+minimize latency due to ``malloc`` and ``free``. Several models, such
+as ``SpinerEOS`` also use the persistency of these arrays to cache
+useful quantities for a performance boost.
+
+EOS Modifiers
+--------------
+
+``EOS`` models can be *modified* by templated classes we call
+*modifiers*. A modifier has exactly the same API as an ``EOS``, but
+provides some internal transformation on inputs and outputs. For
+example the ``ShiftedEOS`` modifier changes the zero point energy of a
```

Comment:
You're concerned about confusing this with vacuum energy? Fair point. Fixed.

Reply:
As I understand the concepts, yes the zero point energy is basically the same concept as the vacuum energy in the sense that both represent the lowest possible _quantized_ state for the system. The difference of course being whether you're in a vacuum or a material and the corresponding change in the underlying shape of the potential.

As an aside, this is the difference between the zero-Kelvin isotherm and the cold curve since the cold curve is just the potential well in which the ions are confined and not the quantized states within the well.

Reply:
Yep second quantization is the same whether you're in vacuum or on a lattice. :+1: